### PR TITLE
Include encoding in signature payload (#30174)

### LINKED ISSUE #30119: Commits which include `encoding` header fail to verify when otherwise valid
### Description

Hello! I'm an engineer on the [Nova](https://nova.app) project, an IDE for the Mac. As a tiny bit of context that becomes important later, we use [libgit2](http://libgit2.org) for our Git integration.

We received a report that a user who utilizes SSH keys for commit signing was seeing those made in Nova as invalid in Gitea. After some diagnosis, we found that the Git CLI was showing them as valid, but not once pushed to Gitea. We were then able to replicate this same behavior in our testing instance of Gitea 1.21.8.

After looking deeper, it seems that the only difference between commits made with the Git CLI and Nova is that those from Nova/libgit2 include an `encoding` header in the commit object, set to `UTF-8`. As far as I've been able to determine, this is the default value from Git's spec, and presumably the Git CLI just never includes it for succinctness. libgit2, however, is including it always.

If we disable inclusion of that header, Gitea will validate the commits just fine, the same as those from the Git CLI.

I am _not versed_ in Go code at all (so please forgive me!) but I think I've traced the issue to somewhere around [this chunk of code](https://github.com/go-gitea/gitea/blob/f47e00d9d3c3bd58b5944a29c4ff5cec0357520a/modules/git/commit_reader.go#L72). I'm guessing that either it's not considering any header name which it doesn't already handle (and thus is dropping them from the payload that is used for validation) or something similar.

If I've glossed over anything important, just let me know, and thank you to the maintainers for your time!

### Gitea Version

1.21.8

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

Screenshot 2024-03-26 at 4 39 59 PM

### Git Version

2.44.0

### Operating System

Linux / Ubuntu in Docker

### How are you running Gitea?

We are running the self-installable Gitea instance running in Docker for testing purposes. I am not sure what our reporting user's instance was running in, or whether it was a self-instance or a Cloud / Enterprise instance. If you need, I can have our support team reach out and ask.

### Database

SQLite

### PR DESCRIPTION
**Pull request by GiteaBot**
Backport #30174 by @KN4CK3R

Fixes #30119

Include the encoding in the signature payload.

before
grafik

after
grafik

### CODE REVIEW

**Review by delvh — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
@KN4CK3R lint failing:

````
Error: modules/git/commit_test.go:152:10: undefined: Sha1Hash (typecheck)

File: modules/git/commit_test.go [hunk 1]
```diff
@@ -125,6 +125,73 @@ empty commit`, commitFromReader.Signature.Payload)
...
+ G8smGxXUeFEt2cP1dj2W8EgAxuA9sTnH9dqI5aRqy5ifDjuya7Emm8sdOUvtGdmn
+ SONRzusmu5n3DgV956REL7x62h7JuqmBz/12HZkr0z0zgXkcZ04q08pSJATX5N1F
+ yN+tWxTsWg+zhDk96d5Esdo9JMjcFvPv0eioo30GAERaz1hoD7zCMT4jgUFTQwgz
+ jw4YcO5u
+ =r3UU
+ -----END PGP SIGNATURE-----
+
+ISO-8859-1`
+
+	sha := &Sha1Hash{0xfe, 0xaf, 0x4b, 0xa6, 0xbc, 0x63, 0x5f, 0xec, 0x44, 0x2f, 0x46, 0xdd, 0xd4, 0x51, 0x24, 0x16, 0xec, 0x43, 0xc2, 0xc2}
```
Comment (KN4CK3R) [inline modules/git/commit_test.go]:
```suggestion
	sha := &Sha1{0xfe, 0xaf, 0x4b, 0xa6, 0xbc, 0x63, 0x5f, 0xec, 0x44, 0x2f, 0x46, 0xdd, 0xd4, 0x51, 0x24, 0x16, 0xec, 0x43, 0xc2, 0xc2}
```

File: modules/git/commit_test.go [hunk 2]
```diff
@@ -125,6 +125,73 @@ empty commit`, commitFromReader.Signature.Payload)
...
+ G8smGxXUeFEt2cP1dj2W8EgAxuA9sTnH9dqI5aRqy5ifDjuya7Emm8sdOUvtGdmn
+ SONRzusmu5n3DgV956REL7x62h7JuqmBz/12HZkr0z0zgXkcZ04q08pSJATX5N1F
+ yN+tWxTsWg+zhDk96d5Esdo9JMjcFvPv0eioo30GAERaz1hoD7zCMT4jgUFTQwgz
+ jw4YcO5u
+ =r3UU
+ -----END PGP SIGNATURE-----
+
+ISO-8859-1`
+
+	sha := &Sha1{0xfe, 0xaf, 0x4b, 0xa6, 0xbc, 0x63, 0x5f, 0xec, 0x44, 0x2f, 0x46, 0xdd, 0xd4, 0x51, 0x24, 0x16, 0xec, 0x43, 0xc2, 0xc2}
```
Comment (KN4CK3R) [inline modules/git/commit_test.go]:
```suggestion
	sha := &SHA1{0xfe, 0xaf, 0x4b, 0xa6, 0xbc, 0x63, 0x5f, 0xec, 0x44, 0x2f, 0x46, 0xdd, 0xd4, 0x51, 0x24, 0x16, 0xec, 0x43, 0xc2, 0xc2}
```

File: modules/git/commit_test.go [hunk 3]
```diff
@@ -125,6 +125,73 @@ empty commit`, commitFromReader.Signature.Payload)
...
+ G8smGxXUeFEt2cP1dj2W8EgAxuA9sTnH9dqI5aRqy5ifDjuya7Emm8sdOUvtGdmn
+ SONRzusmu5n3DgV956REL7x62h7JuqmBz/12HZkr0z0zgXkcZ04q08pSJATX5N1F
+ yN+tWxTsWg+zhDk96d5Esdo9JMjcFvPv0eioo30GAERaz1hoD7zCMT4jgUFTQwgz
+ jw4YcO5u
+ =r3UU
+ -----END PGP SIGNATURE-----
+
+ISO-8859-1`
+
+	sha := &SHA1{0xfe, 0xaf, 0x4b, 0xa6, 0xbc, 0x63, 0x5f, 0xec, 0x44, 0x2f, 0x46, 0xdd, 0xd4, 0x51, 0x24, 0x16, 0xec, 0x43, 0xc2, 0xc2}
```
Comment (KN4CK3R) [inline modules/git/commit_test.go]:
```suggestion
	sha := SHA1{0xfe, 0xaf, 0x4b, 0xa6, 0xbc, 0x63, 0x5f, 0xec, 0x44, 0x2f, 0x46, 0xdd, 0xd4, 0x51, 0x24, 0x16, 0xec, 0x43, 0xc2, 0xc2}
```

**Comment by silverwind** [comment]
@lunny v1.21 branch still expects `mysql5` and `mysql8`. Can this be changed per-branch?

**Comment by lunny** [comment]
> @lunny v1.21 branch still has `mysql5` and `mysql8`. Can this be changed per-branch?

This PR is unrelated to database charset. It's about git signature's text encoding. So I don't know what did you mean.

**Comment by KN4CK3R** [comment]
The pipeline did not run for MySQL.

**Comment by silverwind** [comment]
> > @lunny v1.21 branch still has `mysql5` and `mysql8`. Can this be changed per-branch?
> 
> This PR is unrelated to database charset. It's about git signature's text encoding. So I don't know what did you mean.

I mean this:

image

1.22 and 1.23 have `test-mysql`
1.21 has `test-mysql5` and `test-mysql8`

Please adapt the github protected branches config accordingly.

**Comment by lunny** [comment]
> > > @lunny v1.21 branch still has `mysql5` and `mysql8`. Can this be changed per-branch?
> > 
> > 
> > This PR is unrelated to database charset. It's about git signature's text encoding. So I don't know what did you mean.
> 
> I mean this:
> image
> 
> 1.22 and 1.23 have `test-mysql` 1.21 has `test-mysql5` and `test-mysql8`
> 
> Please adapt the github protected branches config accordingly.

All release branches shared the same rules. I think this PR just need a new push to trigger a new CI to resolve the problem.

**Comment by silverwind** [comment]
> All release branches shared the same rules

You need to make a separate rule for the 1.21 branch. A push won't solve it.

**Pull request merged**